src/transformers/modeling_utils.py

```diff
@@ -996,16 +1076,35 @@ def _generate_beam_search(
                 # Compute next scores
                 next_scores = torch.gather(_scores, -1, next_tokens)  # (batch_size, num_beams * 2)
 
+                # sort the sampled vector to make sure that the first num_beams samples are the best
+                next_scores, next_scores_indices = torch.sort(next_scores, descending=True, dim=1)
+                next_tokens = torch.gather(next_tokens, -1, next_scores_indices)  # (batch_size, num_beams * 2)
             else:
                 # do greedy beam search
                 scores = F.log_softmax(next_token_logits, dim=-1)  # (batch_size * num_beams, vocab_size)
+
+                if is_encoder_decoder:  # TODO(PVP) to be refactored later - do we need this boolean flag here?
+                    #                    scores[scores != scores] = -math.inf  # block nans  => seems very hacky here
+                    #                    scores[:, pad_token_id] = -math.inf => seems very hacky here
+                    # TODO(SS): fairseq also takes out <unk> every step, and has unk at slot 3
+                    #                    if cur_len == 0:  # Force BOS to be chosen  => also very hacky ... seems also to work without this line
```

Comment:
Regarding 3).
Also a hack to force the `<BOS>` to be generated. Seems to work without for the moment - could it work without this hack? @sshleifer @thomwolf @LysandreJik

Reply:
doesn't work without that hack sadly.